File: srcs/server/agents/builtin/llm_openai.go
package builtin

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"time"
)

// OpenAIClient implements LLMClient for OpenAI API.
type OpenAIClient struct {
	APIKey string
	Client *http.Client
}

func NewOpenAIClient(apiKey string) *OpenAIClient {
	return &OpenAIClient{
		APIKey: apiKey,
		Client: &http.Client{Timeout: 2 * time.Minute},
	}
}

func (c *OpenAIClient) Chat(ctx context.Context, req ChatRequest) (ChatResponse, error) {
	// Map our ChatRequest to OpenAI's payload
	type openaiMessage struct {
		Role    string `json:"role"`
		Content string `json:"content"`
	}

	var messages []openaiMessage
	if req.System != "" {
		messages = append(messages, openaiMessage{Role: "system", Content: req.System})
	}

	for _, m := range req.Messages {
		messages = append(messages, openaiMessage{
			Role:    string(m.Role),
			Content: m.Content,
		})
	}

	payload := map[string]interface{}{
		"model":       req.Model,
		"messages":    messages,
	}

	body, _ := json.Marshal(payload)

	httpReq, err := http.NewRequestWithContext(ctx, "POST", "https://api.openai.com/v1/chat/completions", bytes.NewReader(body))
	if err != nil {
		return ChatResponse{}, err
	}

	httpReq.Header.Set("Authorization", "Bearer "+c.APIKey)
	httpReq.Header.Set("Content-Type", "application/json")

	resp, err := c.Client.Do(httpReq)
	if err != nil {
		return ChatResponse{}, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		return ChatResponse{}, fmt.Errorf("openai api error: %s", resp.Status)
	}

	var result struct {
		Choices []struct {
			Message struct {
				Content string `json:"content"`
			} `json:"message"`
		} `json:"choices"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return ChatResponse{}, err
	}

	content := ""
	if len(result.Choices) > 0 {
		content = result.Choices[0].Message.Content
	}

	return ChatResponse{
		Message: Message{
			Role:    RoleAssistant,
			Content: content,
		},
	}, nil
}

// ChatStream implements real SSE streaming for OpenAI.
func (c *OpenAIClient) ChatStream(ctx context.Context, req ChatRequest, chunkChan chan<- ChatResponseChunk) error {
	req.Stream = true
	b, err := json.Marshal(req)
	if err != nil {
		return err
	}

	httpReq, err := http.NewRequestWithContext(ctx, "POST", "https://api.openai.com/v1/chat/completions", bytes.NewReader(b))
	if err != nil {
		return err
	}
	httpReq.Header.Set("Authorization", "Bearer "+c.APIKey)
	httpReq.Header.Set("Content-Type", "application/json")
	httpReq.Header.Set("Accept", "text/event-stream")

	resp, err := c.Client.Do(httpReq)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("openai streaming error: %s", resp.Status)
	}

	scanner := bufio.NewScanner(resp.Body)
	for scanner.Scan() {
		line := scanner.Text()
		if !strings.HasPrefix(line, "data: ") {
			continue
		}
		data := strings.TrimPrefix(line, "data: ")
		if data == "[DONE]" {
			break
		}
		var chunk struct {
			Choices []struct {
				Delta struct {
					Content string `json:"content"`
					ToolCalls []struct {
						ID       string `json:"id,omitempty"`
						Function struct {
							Name      string `json:"name,omitempty"`
							Arguments string `json:"arguments,omitempty"`
						} `json:"function,omitempty"`
					} `json:"tool_calls,omitempty"`
				} `json:"delta"`
				FinishReason string `json:"finish_reason"`
			} `json:"choices"`
		}
		if err := json.Unmarshal([]byte(data), &chunk); err == nil {
			if len(chunk.Choices) > 0 {
				var tcs []ToolCall
				for _, t := range chunk.Choices[0].Delta.ToolCalls {
					tcs = append(tcs, ToolCall{ID: t.ID, Name: t.Function.Name, Arguments: []byte(t.Function.Arguments)})
				}
				chunkChan <- ChatResponseChunk{
					Delta:      chunk.Choices[0].Delta.Content,
					StopReason: chunk.Choices[0].FinishReason,
					ToolCalls:  tcs,
				}
			}
		}
	}
	return scanner.Err()
}

File: srcs/server/agents/builtin/llm_openai.cc
#include "srcs/server/agents/builtin/llm_openai.h"

#include <utility>

#include "srcs/server/agents/builtin/http_client.h"
#include "srcs/server/agents/builtin/llm_parsing.h"
#include "absl/status/status.h"
#include "absl/strings/str_cat.h"
#include "nlohmann/json.hpp"

namespace ohc::agent {

OpenAIClient::OpenAIClient(std::string api_key, std::string base_url)
    : api_key_(std::move(api_key)), base_url_(std::move(base_url)) {}

absl::StatusOr<ChatResponse> OpenAIClient::Chat(const ChatRequest& req) {
  // ---- Build messages array ------------------------------------------------
  nlohmann::json messages = nlohmann::json::array();

  if (!req.system.empty()) {
    messages.push_back({{"role", "system"}, {"content", req.system}});
  }

  for (const auto& m : req.messages) {
    const std::string role(RoleToStringView(m.role));

    // Tool-result messages: OpenAI expects one message per result.
    if (m.role == Role::kTool) {
      for (const auto& tr : m.tool_results) {
        messages.push_back({
            {"role",         "tool"},
            {"tool_call_id", tr.tool_call_id},
            {"content",      tr.content.empty() ? tr.error : tr.content},
        });
      }
      continue;
    }

    nlohmann::json msg = {{"role", role}, {"content", m.content}};

    if (!m.tool_calls.empty()) {
      nlohmann::json tcs = nlohmann::json::array();
      for (const auto& tc : m.tool_calls) {
        tcs.push_back({
            {"id",   tc.id},
            {"type", "function"},
            {"function", {{"name", tc.name},
                          {"arguments", tc.arguments.dump()}}},
        });
      }
      msg["tool_calls"] = std::move(tcs);
    }
    messages.push_back(std::move(msg));
  }

  // ---- Build payload -------------------------------------------------------
  nlohmann::json payload;
  payload["model"]    = req.model;
  payload["messages"] = std::move(messages);
  if (req.max_tokens > 0) payload["max_tokens"]  = req.max_tokens;
  if (req.temperature >= 0.0f) payload["temperature"] = req.temperature;

  if (!req.tool_definitions.empty()) {
    nlohmann::json tools = nlohmann::json::array();
    for (const auto& td : req.tool_definitions) {
      tools.push_back({
          {"type", "function"},
          {"function", {{"name",        td.name},
                        {"description", td.description},
                        {"parameters",  td.parameters}}},
      });
    }
    payload["tools"] = std::move(tools);
  }

  // ---- HTTP POST -----------------------------------------------------------
  HttpRequest http_req;
  http_req.url     = absl::StrCat(base_url_, "/chat/completions");
  http_req.method  = "POST";
  http_req.body    = payload.dump();
  http_req.headers = {{"Authorization", absl::StrCat("Bearer ", api_key_)},
                      {"Content-Type",  "application/json"}};

  auto resp_or = HttpDo(http_req);
  if (!resp_or.ok()) return resp_or.status();
  if (resp_or->status_code != 200) {
    return absl::InternalError(absl::StrCat(
        "OpenAI API HTTP ", resp_or->status_code, ": ", resp_or->body));
  }

  // ---- Parse response ------------------------------------------------------
  return ParseOpenAIChatResponse(resp_or->body);
}

}  // namespace ohc::agent
